botocore/paginate.py

```diff
@@ -323,22 +329,37 @@ def _parse_starting_token(self):
             except ValueError:
                 raise ValueError("Bad starting token: %s" %
                                  self._starting_token)
-        for part in parts:
+        for token_index, part in enumerate(parts):
             if part == 'None':
                 next_token.append(None)
+            elif self._is_dict_token(token_index):
+                try:
+                    # Some services (such as dynamodb) non-string pagination
+                    # tokens, which will fail validation if they aren't
+                    # converted to their proper type.
+                    next_token.append(ast.literal_eval(part))
+                except ValueError:
+                    logger.debug(
+                        "Dict pagination token failed to parse as dict.")
+                    next_token.append(part)
             else:
                 next_token.append(part)
         return next_token, index
 
+    def _is_dict_token(self, token_index):
+        token_type = self._op_model.input_shape.members.get(
+            self._input_token[token_index]).type_name
+        return token_type == 'map' or token_type == 'structure'
 
 
 class Paginator(object):
 
     PAGE_ITERATOR_CLS = PageIterator
 
-    def __init__(self, method, pagination_config):
+    def __init__(self, method, pagination_config, operation_model):
```

Comment:
This reminds me another trick we used [before](https://github.com/boto/botocore/blob/develop/botocore/docs/shape.py#L27). If we define an optional setting like this:

    def __init__(self, method, pagination_config, settings=None):
        self._settings = {} if settings is None else settings

we can then initialize this class in client.py by something like this:

    paginator = documented_paginator_cls(
        ..., ...,
        {"dict_tokens": _find_all_dict_token_names_in_advance()})

Now the `_is_dict_token()` can be implemented as:

    def _is_dict_token(self, token_index):
        return self._input_token[token_index] in self._settings.get('dict_tokens', [])

The benefit is:

* The paginator interface remains backward compatible (although it is not really that important here, because it is not an open API). As a result, we don't need to modify tons of existing test cases.
* The paginator does not couple with the model logic. Should we change model APIs in the future (very unlikely), this paginator will not need to change at all.